In [1]:
# 4/17/2024
# CSC354 – Assignmen3 – ML – Support Vector Machines
# YAHYA IRFAN 
# FA21-BSE-044
# Use the famous ML Iris dataset and only the first two features (input variables). Try to fit an SVM using
# polynomial (degree = 2) and Gaussian (sigma = 1) (keeping the rest of the parameter settings to default)
# kernels on the dataset using an 80/20 split.
# 1. Which kernel settings result in better performance?
# 2. Variate both C and degree (try 3 different combinations) and see how SVM (polynomial) reacts. Report 
# your findings.
# 3. Variate both C and sigma (try 3 different combinations) and see how SVM (Gaussian) reacts. Report 
# your findings.
# Note: To load and play with Iris dataset, following commands may help
# from sklearn import datasets
# iris = datasets.load_iris()
# X = iris.data[:, :2] # use the first two features
# Y = iris.target



from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC

# Load Iris dataset
iris = datasets.load_iris()
X = iris.data[:, :2] 
Y = iris.target

# Split  (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

#(degree = 2)
svm_poly = SVC(kernel='poly', degree=2)
svm_poly.fit(X_train, y_train)
accuracy_poly = svm_poly.score(X_test, y_test)

# (sigma = 1)
svm_rbf = SVC(kernel='rbf', gamma=1)
svm_rbf.fit(X_train, y_train)
accuracy_rbf = svm_rbf.score(X_test, y_test)

print("Accuracy with polynomial kernel:", accuracy_poly)
print("Accuracy with Gaussian kernel:", accuracy_rbf)


C_values = [0.1, 1, 10]
degree_values = [2, 3, 4]

for C in C_values:
    for degree in degree_values:
        svm_poly = SVC(kernel='poly', degree=degree, C=C)
        svm_poly.fit(X_train, y_train)
        accuracy = svm_poly.score(X_test, y_test)
        print("C =", C, ", Degree =", degree, ", Accuracy:", accuracy)

C_values = [0.1, 1, 10]
sigma_values = [0.1, 1, 10]

for C in C_values:
    for sigma in sigma_values:
        svm_rbf = SVC(kernel='rbf', gamma=1/(2*sigma**2), C=C)
        svm_rbf.fit(X_train, y_train)
        accuracy = svm_rbf.score(X_test, y_test)
        print("C =", C, ", Sigma =", sigma, ", Accuracy:", accuracy)

Accuracy with polynomial kernel: 0.8333333333333334
Accuracy with Gaussian kernel: 0.9
C = 0.1 , Degree = 2 , Accuracy: 0.9
C = 0.1 , Degree = 3 , Accuracy: 0.8333333333333334
C = 0.1 , Degree = 4 , Accuracy: 0.8
C = 1 , Degree = 2 , Accuracy: 0.8333333333333334
C = 1 , Degree = 3 , Accuracy: 0.8333333333333334
C = 1 , Degree = 4 , Accuracy: 0.8
C = 10 , Degree = 2 , Accuracy: 0.9
C = 10 , Degree = 3 , Accuracy: 0.8
C = 10 , Degree = 4 , Accuracy: 0.8
C = 0.1 , Sigma = 0.1 , Accuracy: 0.3
C = 0.1 , Sigma = 1 , Accuracy: 0.8333333333333334
C = 0.1 , Sigma = 10 , Accuracy: 0.3
C = 1 , Sigma = 0.1 , Accuracy: 0.8333333333333334
C = 1 , Sigma = 1 , Accuracy: 0.9
C = 1 , Sigma = 10 , Accuracy: 0.3
C = 10 , Sigma = 0.1 , Accuracy: 0.6666666666666666
C = 10 , Sigma = 1 , Accuracy: 0.9
C = 10 , Sigma = 10 , Accuracy: 0.9


Answers:

1. Which kernel settings result in better performance? 
Accuracy with polynomial kernel: 0.8333333333333334
Accuracy with Gaussian kernel: 0.9
The Gaussian kernel resulted in better performance with a slightly higher accuracy of 0.9 compared to the polynomial kernel's accuracy of 0.8333333333333334.


2. Variating both C and degree for the SVM with polynomial kernel:

Combination 1: C = 0.1, Degree = 2, Accuracy: 0.9
Combination 2: C = 1, Degree = 3, Accuracy: 0.8333333333333334
Combination 3: C = 10, Degree = 2, Accuracy: 0.9
We observe that the SVM with polynomial kernel achieves the highest accuracy of 0.9 when C = 0.1 and Degree = 2.


3. Variating both C and sigma for the SVM with Gaussian kernel:

Combination 1: C = 0.1, Sigma = 0.1, Accuracy: 0.3
Combination 2: C = 1, Sigma = 1, Accuracy: 0.9
Combination 3: C = 10, Sigma = 1, Accuracy: 0.9
Here, we can see that for both combinations where C = 1 and C = 10 with Sigma = 1, the SVM with Gaussian kernel achieves the highest accuracy of 0.9.  